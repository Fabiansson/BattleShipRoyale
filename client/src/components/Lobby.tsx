import React, { useState, useContext } from "react";
import Grid from '@material-ui/core/Grid';
import Paper from '@material-ui/core/Paper';
import FormControl from '@material-ui/core/FormControl';
import InputLabel from '@material-ui/core/InputLabel';
import Select from '@material-ui/core/Select';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import Switch from '@material-ui/core/Switch';
import MenuItem from '@material-ui/core/MenuItem';
import Avatar from '@material-ui/core/Avatar';
import Button from '@material-ui/core/Button';

import SocketContext from '../services/SocketProvider';
import UserContext from '../services/UserProvider';
import { createStyles, Theme } from "@material-ui/core";
import { makeStyles } from "@material-ui/core/styles";
import { GeneralGameState } from "../App";
import Chat from "./Chat";

export interface LobbyProps {
    generalGameState: GeneralGameState;
}

export interface GameSettings {
    privateLobby: boolean,
    rounds: number
}

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        list: {
            paddingInlineStart: '0'
        },
        avatar: {
            width: theme.spacing(7),
            height: theme.spacing(7),
        },
        playerListElement: {
            float: 'left',
            listStyleType: 'none'
        }
    }),
);

function Lobby(props: LobbyProps) {
    const classes = useStyles();
    let initialGameSettings: GameSettings = {
        privateLobby: true,
        rounds: 5
    }
    const [gameSettings, setGameSettings] = useState<GameSettings>(initialGameSettings);
    const userId = useContext(UserContext);
    const socket = useContext(SocketContext);

    const handlePrivateChange = () => {
        let newGameSettings: GameSettings = { ...gameSettings, privateLobby: !gameSettings.privateLobby };
        setGameSettings(newGameSettings);
        socket?.emit('gameSettings', newGameSettings);
    }

    const handleRoundOptionChange = (event: any) => {
        let newGameSettings: GameSettings = { ...gameSettings, rounds: event.target.value };
        setGameSettings(newGameSettings);
        socket?.emit('gameSettings', newGameSettings);
    }

    const startGame = () => {
        socket?.emit('startGame');
    }

    return (
        <React.Fragment>

            <Grid
                container
                spacing={0}
                direction="column"
                alignItems="center"
                justify="center"
                style={{ minHeight: '100vh' }}
            >

                <Grid item xs={6}>
                    <Paper elevation={3} style={{ padding: '3em' }}>
                        Invite Link: <a href={'http://localhost:3000/' + props.generalGameState.gameId}>{'http://localhost:3000/' + props.generalGameState.gameId}</a>
                        <Grid container spacing={2}>
                            {props.generalGameState.admin === userId &&
                                <Grid item xs={6}>
                                    <h2>Settings</h2>
                                    <FormControl variant="outlined" >
                                        <InputLabel id="demo-simple-select-outlined-label">Rounds</InputLabel>
                                        <Select value={gameSettings.rounds} onChange={handleRoundOptionChange} label="Rounds">
                                            <MenuItem value={1}>1</MenuItem>
                                            <MenuItem value={5}>5</MenuItem>
                                            <MenuItem value={8}>8</MenuItem>
                                            <MenuItem value={10}>10</MenuItem>
                                            <MenuItem value={15}>15</MenuItem>
                                        </Select>
                                    </FormControl>
                                    <FormControlLabel
                                        control={<Switch checked={gameSettings.privateLobby} onChange={handlePrivateChange} name="PrivateLobby" color="primary" />}
                                        label="Private Lobby"
                                        labelPlacement="top"
                                    />
                                </Grid>}
                            <Grid item xs={6}>
                                <h2>Players</h2>
                                <ul className={classes.list}>
                                    {props.generalGameState.playerNames.map((player, i) =>
                                        <li key={i} className={classes.playerListElement}>
                                            <div style={{textAlign: 'center'}}><Avatar className={classes.avatar} src={'https://avatars.dicebear.com/v2/avataaars/' + player + 'd.svg'}>{player}</Avatar>
                                            <p>{player}</p></div>
                                            
                                        </li>)}
                                </ul>
                            </Grid>
                            {props.generalGameState.players.length >= 1 && props.generalGameState.admin === userId &&
                                <Grid item xs={12}>
                                    <Button id="startButton" variant="contained" color="primary" onClick={startGame}>Start</Button>
                                </Grid>}
                            <Grid item xs={12}>
                                <Chat />
                            </Grid>
                        </Grid>
                    </Paper>
                </Grid>
            </Grid>
        </React.Fragment>
    );
}

export default Lobby;
