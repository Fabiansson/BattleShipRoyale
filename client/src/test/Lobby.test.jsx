import React from 'react';

import Lobby from '../components/Lobby';
import renderer from 'react-test-renderer';


describe('Lobby', () => {
    it('renders correctly', () => {
        const generalGameStateMock = {
            gameId: 'adasdad',
            admin: 'Peter',
            players: [{
                playerId: 'asdasda',
                playerName: 'afffff'
            }]
        }

        const component = renderer.create(<Lobby generalGameState={generalGameStateMock}/>)

        expect(component).toMatchSnapshot();
    });
});