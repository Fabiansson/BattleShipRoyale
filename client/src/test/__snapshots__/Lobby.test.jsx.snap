// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lobby renders correctly 1`] = `
<div
  className="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
  style={
    Object {
      "minHeight": "100vh",
    }
  }
>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
    style={
      Object {
        "width": "50%",
      }
    }
  >
    <div
      className="MuiPaper-root MuiPaper-elevation3 MuiPaper-rounded"
      style={
        Object {
          "padding": "3em",
        }
      }
    >
      Invite Link: 
      <a
        href="http://localhost:3000/adasdadarschloch"
      >
        http://localhost:3000/adasdad
      </a>
      <div
        className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
      >
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
        >
          <h2>
            Players
          </h2>
          <ul
            className="makeStyles-list-1"
          >
            <li
              className="makeStyles-playerListElement-3"
            >
              <div
                style={
                  Object {
                    "textAlign": "center",
                  }
                }
              >
                <div
                  className="MuiAvatar-root MuiAvatar-circle makeStyles-avatar-2"
                >
                  <img
                    className="MuiAvatar-img"
                    src="https://avatars.dicebear.com/v2/avataaars/afffffd.svg"
                  />
                </div>
                <p>
                  afffff
                </p>
              </div>
            </li>
          </ul>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          style={
            Object {
              "height": "300px",
            }
          }
        >
          <div
            className="makeStyles-rootGrid-142"
          >
            <div
              className="makeStyles-listDiv-144"
              id="listDiv"
            >
              <ul
                className="makeStyles-list-143"
              />
              <div />
            </div>
            <div
              className="makeStyles-inputField-146"
            >
              <div
                className="MuiInputBase-root MuiInput-root MuiInput-underline"
                onClick={[Function]}
              >
                <input
                  aria-label="description"
                  className="MuiInputBase-input MuiInput-input"
                  id="submitText"
                  onAnimationStart={[Function]}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  placeholder="Message"
                  type="text"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
